This patches gives 3Fs, unsure why but it seems we are losing some inference info for diagnostics ...

```
failures:

---- [ui] tests/ui/lazy-type-alias-impl-trait/branches3.rs stdout ----
diff of stderr:

1	error[E0282]: type annotations needed
-	  --> $DIR/branches3.rs:8:10
+	  --> $DIR/branches3.rs:8:13
3	   |
4	LL |         |s| s.len()
-	   |          ^  - type must be known at this point
-	   |
-	help: consider giving this closure parameter an explicit type
-	   |
-	LL |         |s: /* Type */| s.len()
-	   |           ++++++++++++
+	   |             ^ cannot infer type
11	
12	error[E0282]: type annotations needed
-	  --> $DIR/branches3.rs:15:10
+	  --> $DIR/branches3.rs:15:13
14	   |
15	LL |         |s| s.len()
-	   |          ^  - type must be known at this point
-	   |
-	help: consider giving this closure parameter an explicit type
-	   |
-	LL |         |s: /* Type */| s.len()
-	   |           ++++++++++++
+	   |             ^ cannot infer type
22	
23	error[E0282]: type annotations needed
-	  --> $DIR/branches3.rs:23:10
+	  --> $DIR/branches3.rs:23:13
25	   |
26	LL |         |s| s.len()
-	   |          ^  - type must be known at this point
-	   |
-	help: consider giving this closure parameter an explicit type
-	   |
-	LL |         |s: /* Type */| s.len()
-	   |           ++++++++++++
+	   |             ^ cannot infer type
33	
34	error[E0282]: type annotations needed
-	  --> $DIR/branches3.rs:30:10
+	  --> $DIR/branches3.rs:30:13
36	   |
37	LL |         |s| s.len()
-	   |          ^  - type must be known at this point
-	   |
-	help: consider giving this closure parameter an explicit type
-	   |
-	LL |         |s: /* Type */| s.len()
-	   |           ++++++++++++
+	   |             ^ cannot infer type
44	
45	error: aborting due to 4 previous errors
46	


The actual stderr differed from the expected stderr.
Actual stderr saved to /home/gh-spastorino/rust3/build/aarch64-unknown-linux-gnu/test/ui/lazy-type-alias-impl-trait/branches3/branches3.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args lazy-type-alias-impl-trait/branches3.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/home/gh-spastorino/rust3/build/aarch64-unknown-linux-gnu/stage1/bin/rustc" "/home/gh-spastorino/rust3/tests/ui/lazy-type-alias-impl-trait/branches3.rs" "-Zthreads=1" "--target=aarch64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--remap-path-prefix=/home/gh-spastorino/rust3/tests/ui=fake-test-src-base" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/home/gh-spastorino/rust3/build/aarch64-unknown-linux-gnu/test/ui/lazy-type-alias-impl-trait/branches3" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/home/gh-spastorino/rust3/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/home/gh-spastorino/rust3/build/aarch64-unknown-linux-gnu/test/ui/lazy-type-alias-impl-trait/branches3/auxiliary"
stdout: none
--- stderr -------------------------------
error[E0282]: type annotations needed
  --> fake-test-src-base/lazy-type-alias-impl-trait/branches3.rs:8:13
   |
LL |         |s| s.len() //~ ERROR type annotations needed
   |             ^ cannot infer type

error[E0282]: type annotations needed
  --> fake-test-src-base/lazy-type-alias-impl-trait/branches3.rs:15:13
   |
LL |         |s| s.len() //~ ERROR type annotations needed
   |             ^ cannot infer type

error[E0282]: type annotations needed
  --> fake-test-src-base/lazy-type-alias-impl-trait/branches3.rs:23:13
   |
LL |         |s| s.len() //~ ERROR type annotations needed
   |             ^ cannot infer type

error[E0282]: type annotations needed
  --> fake-test-src-base/lazy-type-alias-impl-trait/branches3.rs:30:13
   |
LL |         |s| s.len() //~ ERROR type annotations needed
   |             ^ cannot infer type

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0282`.
------------------------------------------


---- [ui] tests/ui/unboxed-closures/unboxed-closures-failed-recursive-fn-2.rs stdout ----
diff of stderr:

-	error[E0282]: type annotations needed for `Option<T>`
-	  --> $DIR/unboxed-closures-failed-recursive-fn-2.rs:8:9
+	error[E0282]: type annotations needed
+	  --> $DIR/unboxed-closures-failed-recursive-fn-2.rs:17:32
3	   |
-	LL |     let mut closure0 = None;
-	   |         ^^^^^^^^^^^^
-	...
7	LL |                         return c();
-	   |                                --- type must be known at this point
-	   |
-	help: consider giving `closure0` an explicit type, where the placeholders `_` are specified
-	   |
-	LL |     let mut closure0: Option<T> = None;
-	   |                     +++++++++++
+	   |                                ^^^ cannot infer type
14	
15	error: aborting due to previous error
16	


The actual stderr differed from the expected stderr.
Actual stderr saved to /home/gh-spastorino/rust3/build/aarch64-unknown-linux-gnu/test/ui/unboxed-closures/unboxed-closures-failed-recursive-fn-2/unboxed-closures-failed-recursive-fn-2.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args unboxed-closures/unboxed-closures-failed-recursive-fn-2.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/home/gh-spastorino/rust3/build/aarch64-unknown-linux-gnu/stage1/bin/rustc" "/home/gh-spastorino/rust3/tests/ui/unboxed-closures/unboxed-closures-failed-recursive-fn-2.rs" "-Zthreads=1" "--target=aarch64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--remap-path-prefix=/home/gh-spastorino/rust3/tests/ui=fake-test-src-base" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/home/gh-spastorino/rust3/build/aarch64-unknown-linux-gnu/test/ui/unboxed-closures/unboxed-closures-failed-recursive-fn-2" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/home/gh-spastorino/rust3/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/home/gh-spastorino/rust3/build/aarch64-unknown-linux-gnu/test/ui/unboxed-closures/unboxed-closures-failed-recursive-fn-2/auxiliary"
stdout: none
--- stderr -------------------------------
error[E0282]: type annotations needed
  --> fake-test-src-base/unboxed-closures/unboxed-closures-failed-recursive-fn-2.rs:17:32
   |
LL |                         return c();
   |                                ^^^ cannot infer type

error: aborting due to previous error

For more information about this error, try `rustc --explain E0282`.
------------------------------------------


---- [ui] tests/ui/type-alias-impl-trait/closures_in_branches.rs stdout ----
diff of stderr:

1	error[E0282]: type annotations needed
-	  --> $DIR/closures_in_branches.rs:7:10
+	  --> $DIR/closures_in_branches.rs:7:13
3	   |
4	LL |         |x| x.len()
-	   |          ^  - type must be known at this point
-	   |
-	help: consider giving this closure parameter an explicit type
-	   |
-	LL |         |x: /* Type */| x.len()
-	   |           ++++++++++++
+	   |             ^ cannot infer type
11	
12	error[E0282]: type annotations needed
-	  --> $DIR/closures_in_branches.rs:21:10
+	  --> $DIR/closures_in_branches.rs:21:13
14	   |
15	LL |         |x| x.len()
-	   |          ^  - type must be known at this point
-	   |
-	help: consider giving this closure parameter an explicit type
-	   |
-	LL |         |x: /* Type */| x.len()
-	   |           ++++++++++++
+	   |             ^ cannot infer type
22	
23	error: aborting due to 2 previous errors
24	


The actual stderr differed from the expected stderr.
Actual stderr saved to /home/gh-spastorino/rust3/build/aarch64-unknown-linux-gnu/test/ui/type-alias-impl-trait/closures_in_branches/closures_in_branches.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args type-alias-impl-trait/closures_in_branches.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/home/gh-spastorino/rust3/build/aarch64-unknown-linux-gnu/stage1/bin/rustc" "/home/gh-spastorino/rust3/tests/ui/type-alias-impl-trait/closures_in_branches.rs" "-Zthreads=1" "--target=aarch64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--remap-path-prefix=/home/gh-spastorino/rust3/tests/ui=fake-test-src-base" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/home/gh-spastorino/rust3/build/aarch64-unknown-linux-gnu/test/ui/type-alias-impl-trait/closures_in_branches" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/home/gh-spastorino/rust3/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/home/gh-spastorino/rust3/build/aarch64-unknown-linux-gnu/test/ui/type-alias-impl-trait/closures_in_branches/auxiliary"
stdout: none
--- stderr -------------------------------
error[E0282]: type annotations needed
  --> fake-test-src-base/type-alias-impl-trait/closures_in_branches.rs:7:13
   |
LL |         |x| x.len() //~ ERROR type annotations needed
   |             ^ cannot infer type

error[E0282]: type annotations needed
  --> fake-test-src-base/type-alias-impl-trait/closures_in_branches.rs:21:13
   |
LL |         |x| x.len() //~ ERROR type annotations needed
   |             ^ cannot infer type

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0282`.
------------------------------------------



failures:
    [ui] tests/ui/lazy-type-alias-impl-trait/branches3.rs
    [ui] tests/ui/type-alias-impl-trait/closures_in_branches.rs
    [ui] tests/ui/unboxed-closures/unboxed-closures-failed-recursive-fn-2.rs

test result: FAILED. 4 passed; 3 failed; 14577 ignored; 0 measured; 0 filtered out; finished in 0.19s
```

--- a/compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs
+++ b/compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs
@@ -581,7 +581,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
 
         if !errors.is_empty() {
             self.adjust_fulfillment_errors_for_expr_obligation(&mut errors);
-            self.err_ctxt().report_fulfillment_errors(&errors, self.inh.body_id);
+            self.err_ctxt().report_fulfillment_errors(&errors, Some(self.inh.body_def_id));
         }
     }
 
@@ -594,7 +594,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         if !result.is_empty() {
             mutate_fulfillment_errors(&mut result);
             self.adjust_fulfillment_errors_for_expr_obligation(&mut result);
-            self.err_ctxt().report_fulfillment_errors(&result, self.inh.body_id);
+            self.err_ctxt().report_fulfillment_errors(&result, Some(self.inh.body_def_id));
         }
     }
 
@@ -1411,7 +1411,13 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         } else {
             let e = self.tainted_by_errors().unwrap_or_else(|| {
                 self.err_ctxt()
-                    .emit_inference_failure_err((**self).body_id, sp, ty.into(), E0282, true)
+                    .emit_inference_failure_err(
+                        Some(self.inh.body_def_id),
+                        sp,
+                        ty.into(),
+                        E0282,
+                        true,
+                    )
                     .emit()
             });
             let err = self.tcx.ty_error(e);

--- a/compiler/rustc_hir_typeck/src/inherited.rs
+++ b/compiler/rustc_hir_typeck/src/inherited.rs
@@ -58,7 +58,7 @@ pub struct Inherited<'tcx> {
     pub(super) deferred_generator_interiors:
         RefCell<Vec<(LocalDefId, hir::BodyId, Ty<'tcx>, hir::GeneratorKind)>>,
 
-    pub(super) body_id: Option<hir::BodyId>,
+    pub(super) body_def_id: LocalDefId,
 
     /// Whenever we introduce an adjustment from `!` into a type variable,
     /// we record that type variable here. This is later used to inform
@@ -116,7 +116,6 @@ impl<'tcx> Inherited<'tcx> {
         typeck_results: RefCell<ty::TypeckResults<'tcx>>,
     ) -> Self {
         let tcx = infcx.tcx;
-        let body_id = tcx.hir().maybe_body_owned_by(def_id);
 
         Inherited {
             typeck_results,
@@ -130,7 +129,7 @@ impl<'tcx> Inherited<'tcx> {
             deferred_asm_checks: RefCell::new(Vec::new()),
             deferred_generator_interiors: RefCell::new(Vec::new()),
             diverging_type_vars: RefCell::new(Default::default()),
-            body_id,
+            body_def_id: def_id,
             infer_var_info: RefCell::new(Default::default()),
         }
     }

--- a/compiler/rustc_hir_typeck/src/writeback.rs
+++ b/compiler/rustc_hir_typeck/src/writeback.rs
@@ -748,7 +748,7 @@ impl<'cx, 'tcx> Resolver<'cx, 'tcx> {
                 .infcx
                 .err_ctxt()
                 .emit_inference_failure_err(
-                    Some(self.body.id()),
+                    Some(self.tcx.hir().body_owner_def_id(self.body.id())),
                     self.span.to_span(self.tcx),
                     p.into(),
                     E0282,

--- a/compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs
+++ b/compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs
@@ -10,7 +10,7 @@ use rustc_errors::{DiagnosticBuilder, ErrorGuaranteed, IntoDiagnosticArg};
 use rustc_hir as hir;
 use rustc_hir::def::Res;
 use rustc_hir::def::{CtorOf, DefKind, Namespace};
-use rustc_hir::def_id::DefId;
+use rustc_hir::def_id::{DefId, LocalDefId};
 use rustc_hir::intravisit::{self, Visitor};
 use rustc_hir::{Body, Closure, Expr, ExprKind, FnRetTy, HirId, Local, LocalSource};
 use rustc_middle::hir::nested_filter;
@@ -386,7 +386,7 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
     #[instrument(level = "debug", skip(self, error_code))]
     pub fn emit_inference_failure_err(
         &self,
-        body_id: Option<hir::BodyId>,
+        body_def_id: Option<LocalDefId>,
         failure_span: Span,
         arg: GenericArg<'tcx>,
         error_code: TypeAnnotationNeeded,
@@ -403,8 +403,10 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
         };
 
         let mut local_visitor = FindInferSourceVisitor::new(&self, typeck_results, arg);
-        if let Some(body_id) = body_id {
-            let expr = self.tcx.hir().expect_expr(body_id.hir_id);
+        if let Some(body_def_id) = body_def_id
+            && let Some(body_id) = self.tcx.hir().maybe_body_owned_by(body_def_id)
+        {
+            let expr = self.tcx.hir().body(body_id).value;
             local_visitor.visit_expr(expr);
         }
 

--- a/compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs
+++ b/compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs
@@ -24,11 +24,9 @@ use rustc_errors::{
 };
 use rustc_hir as hir;
 use rustc_hir::def::Namespace;
-use rustc_hir::def_id::DefId;
+use rustc_hir::def_id::{DefId, LocalDefId};
 use rustc_hir::intravisit::Visitor;
-use rustc_hir::GenericParam;
-use rustc_hir::Item;
-use rustc_hir::Node;
+use rustc_hir::{GenericParam, Item, Node};
 use rustc_infer::infer::error_reporting::TypeErrCtxt;
 use rustc_infer::infer::{InferOk, TypeTrace};
 use rustc_middle::traits::select::OverflowError;
@@ -129,7 +127,7 @@ pub trait TypeErrCtxtExt<'tcx> {
     fn report_fulfillment_errors(
         &self,
         errors: &[FulfillmentError<'tcx>],
-        body_id: Option<hir::BodyId>,
+        body_def_id: Option<LocalDefId>,
     ) -> ErrorGuaranteed;
 
     fn report_overflow_obligation<T>(
@@ -391,7 +389,7 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
     fn report_fulfillment_errors(
         &self,
         errors: &[FulfillmentError<'tcx>],
-        body_id: Option<hir::BodyId>,
+        body_def_id: Option<LocalDefId>,
     ) -> ErrorGuaranteed {
         #[derive(Debug)]
         struct ErrorDescriptor<'tcx> {
@@ -469,7 +467,7 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
         for from_expansion in [false, true] {
             for (error, suppressed) in iter::zip(errors, &is_suppressed) {
                 if !suppressed && error.obligation.cause.span.from_expansion() == from_expansion {
-                    self.report_fulfillment_error(error, body_id);
+                    self.report_fulfillment_error(error, body_def_id);
                 }
             }
         }
@@ -955,8 +953,7 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                             );
                         }
 
-                        let body_hir_id =
-                            self.tcx.hir().local_def_id_to_hir_id(obligation.cause.body_id);
+                        let body_def_id = obligation.cause.body_id;
                         // Try to report a help message
                         if is_fn_trait
                             && let Ok((implemented_kind, params)) = self.type_implements_fn_trait(
@@ -1037,7 +1034,7 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                             if !self.report_similar_impl_candidates(
                                 impl_candidates,
                                 trait_ref,
-                                body_hir_id,
+                                body_def_id,
                                 &mut err,
                                 true,
                             ) {
@@ -1073,7 +1070,7 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                                     self.report_similar_impl_candidates(
                                         impl_candidates,
                                         trait_ref,
-                                        body_hir_id,
+                                        body_def_id,
                                         &mut err,
                                         true,
                                     );
@@ -1497,7 +1494,7 @@ trait InferCtxtPrivExt<'tcx> {
     fn report_fulfillment_error(
         &self,
         error: &FulfillmentError<'tcx>,
-        body_id: Option<hir::BodyId>,
+        body_def_id: Option<LocalDefId>,
     );
 
     fn report_projection_error(
@@ -1531,7 +1528,7 @@ trait InferCtxtPrivExt<'tcx> {
         &self,
         impl_candidates: Vec<ImplCandidate<'tcx>>,
         trait_ref: ty::PolyTraitRef<'tcx>,
-        body_id: hir::HirId,
+        body_def_id: LocalDefId,
         err: &mut Diagnostic,
         other: bool,
     ) -> bool;
@@ -1564,7 +1561,7 @@ trait InferCtxtPrivExt<'tcx> {
     fn maybe_report_ambiguity(
         &self,
         obligation: &PredicateObligation<'tcx>,
-        body_id: Option<hir::BodyId>,
+        body_def_id: Option<LocalDefId>,
     );
 
     fn predicate_can_apply(
@@ -1650,7 +1647,7 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
     fn report_fulfillment_error(
         &self,
         error: &FulfillmentError<'tcx>,
-        body_id: Option<hir::BodyId>,
+        body_def_id: Option<LocalDefId>,
     ) {
         match error.code {
             FulfillmentErrorCode::CodeSelectionError(ref selection_error) => {
@@ -1664,7 +1661,7 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                 self.report_projection_error(&error.obligation, e);
             }
             FulfillmentErrorCode::CodeAmbiguity => {
-                self.maybe_report_ambiguity(&error.obligation, body_id);
+                self.maybe_report_ambiguity(&error.obligation, body_def_id);
             }
             FulfillmentErrorCode::CodeSubtypeError(ref expected_found, ref err) => {
                 self.report_mismatched_types(
@@ -2029,7 +2026,7 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
         &self,
         impl_candidates: Vec<ImplCandidate<'tcx>>,
         trait_ref: ty::PolyTraitRef<'tcx>,
-        body_id: hir::HirId,
+        body_def_id: LocalDefId,
         err: &mut Diagnostic,
         other: bool,
     ) -> bool {
@@ -2120,9 +2117,7 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                         // FIXME(compiler-errors): This could be generalized, both to
                         // be more granular, and probably look past other `#[fundamental]`
                         // types, too.
-                        self.tcx
-                            .visibility(def.did())
-                            .is_accessible_from(body_id.owner.def_id, self.tcx)
+                        self.tcx.visibility(def.did()).is_accessible_from(body_def_id, self.tcx)
                     } else {
                         true
                     }
@@ -2234,7 +2229,7 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
     fn maybe_report_ambiguity(
         &self,
         obligation: &PredicateObligation<'tcx>,
-        body_id: Option<hir::BodyId>,
+        body_def_id: Option<LocalDefId>,
     ) {
         // Unable to successfully determine, probably means
         // insufficient type information, but could mean
@@ -2277,7 +2272,7 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                 if self.tcx.lang_items().sized_trait() == Some(trait_ref.def_id()) {
                     if let None = self.tainted_by_errors() {
                         self.emit_inference_failure_err(
-                            body_id,
+                            body_def_id,
                             span,
                             trait_ref.self_ty().skip_binder().into(),
                             ErrorCode::E0282,
@@ -2304,7 +2299,13 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                 let subst = data.trait_ref.substs.iter().find(|s| s.has_non_region_infer());
 
                 let mut err = if let Some(subst) = subst {
-                    self.emit_inference_failure_err(body_id, span, subst, ErrorCode::E0283, true)
+                    self.emit_inference_failure_err(
+                        body_def_id,
+                        span,
+                        subst,
+                        ErrorCode::E0283,
+                        true,
+                    )
                 } else {
                     struct_span_err!(
                         self.tcx.sess,
@@ -2348,12 +2349,10 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                                 predicate.to_opt_poly_trait_pred().unwrap(),
                             );
                             if impl_candidates.len() < 10 {
-                                let hir =
-                                    self.tcx.hir().local_def_id_to_hir_id(obligation.cause.body_id);
                                 self.report_similar_impl_candidates(
                                     impl_candidates,
                                     trait_ref,
-                                    body_id.map(|id| id.hir_id).unwrap_or(hir),
+                                    body_def_id.unwrap_or(obligation.cause.body_id),
                                     &mut err,
                                     false,
                                 );
@@ -2375,9 +2374,10 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                     self.suggest_fully_qualified_path(&mut err, def_id, span, trait_ref.def_id());
                 }
 
-                if let (Some(body_id), Some(ty::subst::GenericArgKind::Type(_))) =
-                    (body_id, subst.map(|subst| subst.unpack()))
+                if let (Some(body_def_id), Some(ty::subst::GenericArgKind::Type(_))) =
+                    (body_def_id, subst.map(|subst| subst.unpack()))
                 {
+                    let body_id = self.tcx.hir().body_owned_by(body_def_id);
                     let mut expr_finder = FindExprBySpan::new(span);
                     expr_finder.visit_expr(&self.tcx.hir().body(body_id).value);
 
@@ -2473,7 +2473,7 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                     return;
                 }
 
-                self.emit_inference_failure_err(body_id, span, arg, ErrorCode::E0282, false)
+                self.emit_inference_failure_err(body_def_id, span, arg, ErrorCode::E0282, false)
             }
 
             ty::PredicateKind::Subtype(data) => {
@@ -2487,7 +2487,7 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                 let SubtypePredicate { a_is_expected: _, a, b } = data;
                 // both must be type variables, or the other would've been instantiated
                 assert!(a.is_ty_var() && b.is_ty_var());
-                self.emit_inference_failure_err(body_id, span, a.into(), ErrorCode::E0282, true)
+                self.emit_inference_failure_err(body_def_id, span, a.into(), ErrorCode::E0282, true)
             }
             ty::PredicateKind::Clause(ty::Clause::Projection(data)) => {
                 if predicate.references_error() || self.tainted_by_errors().is_some() {
@@ -2501,7 +2501,7 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                     .find(|g| g.has_non_region_infer());
                 if let Some(subst) = subst {
                     let mut err = self.emit_inference_failure_err(
-                        body_id,
+                        body_def_id,
                         span,
                         subst,
                         ErrorCode::E0284,
@@ -2530,7 +2530,7 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                 let subst = data.walk().find(|g| g.is_non_region_infer());
                 if let Some(subst) = subst {
                     let err = self.emit_inference_failure_err(
-                        body_id,
+                        body_def_id,
                         span,
                         subst,
                         ErrorCode::E0284,

